cachi2/core/package_managers/yarn/resolver.py

```diff
@@ -124,44 +141,217 @@ def resolve_packages(source_dir: RootedPath) -> list[Package]:
+            if (
+                package.cache_path
+                # yarn info seems to always report the cache path for patch dependencies,
+                # but the path doesn't always exist
+                and (cache_path := self._cache_path_as_rooted(package.cache_path)).path.exists()
+            ):
+                log_for_locator("reading package name from %s", cache_path.subpath_from_root)
+                name = self._read_name_from_cache(cache_path)
+            elif orig_package := self._package_mapping.get(locator.package):
+                log_for_locator("resolving the name of the original package")
+                name = self._resolve_package(orig_package).name
+            else:
+                raise _CouldNotResolve(
+                    "the 'yarn info' output does not include either an existing zip archive "
+                    "or the original unpatched package",
+                )
+            version = package.version
+        else:
+            assert_never(locator)
+
+        return _ResolvedPackage(locator, name, version, package.raw_locator)
+
+    def _read_name_from_cache(self, cache_path: RootedPath) -> str:
+        with zipfile.ZipFile(cache_path) as zf:
+            packjson_paths = (
+                filename
+                for filename in zf.namelist()
+                # node_modules/@scope/name/package.json
+                # node_modules/name/package.json
+                if (path := Path(filename)).parts[0] == "node_modules"
+                and len(path.parts) in (3, 4)
+                and path.parts[-1] == "package.json"
+            )
+            packjson_path = next(packjson_paths, None)
+            if packjson_path is None:
+                raise _CouldNotResolve(f"{cache_path.subpath_from_root}: no package.json")
+
+            packjson_content = zf.read(packjson_path)
+
+        try:
+            packjson = json.loads(packjson_content)
+        except json.JSONDecodeError as e:
+            raise _CouldNotResolve(
+                f"{cache_path.subpath_from_root}::{packjson_path}: invalid JSON ({e})"
+            ) from e
+
+        if not (name := packjson.get("name")):
+            raise _CouldNotResolve(
+                f"{cache_path.subpath_from_root}::{packjson_path}: no 'name' attribute"
+            )
+
+        return name
+
+    def _read_name_version_from_packjson(
+        self, packjson_path: RootedPath
+    ) -> tuple[str, Optional[str]]:
+        try:
+            packjson = json.loads(packjson_path.path.read_text())
+        except FileNotFoundError as e:
+            raise _CouldNotResolve(f"missing {packjson_path.subpath_from_root}") from e
+        except json.JSONDecodeError as e:
+            raise _CouldNotResolve(f"{packjson_path.subpath_from_root}: invalid JSON ({e})") from e
+
+        if not (name := packjson.get("name")):
+            raise _CouldNotResolve(f"{packjson_path.subpath_from_root}: no 'name' attribute")
+
+        return name, packjson.get("version")
```

Comment:
So we're assuming that, if version does not exist, we'll just leave it empty? This can potentially be the case of the main package, since it is treated as a workspace.

I'm just wondering if we want to use some approach involving the commit hash for such cases (similar to gomod). But pseudoversions are not a thing for npm/yarn, right?

Reply:
> So we're assuming that, if version does not exist, we'll just leave it empty? This can potentially be the case of the main package, since it is treated as a workspace.

Yeah. It can be the case for the main package and pretty much any dependency other than npm dependencies.

> I'm just wondering if we want to use some approach involving the commit hash for such cases (similar to gomod). But pseudoversions are not a thing for npm/yarn, right?

In the purl yes, in the version no. I don't see a reason to make up versions for packages that don't have one

Reply:
The name and version should be those that make sense within the package manager's ecosystem. If a CVE is found in `some-package@1.2.3`, it should be possible to identify that lib in our SBOMs regardless of where the package came from (false positives possible of course).

The purl should then make it possible to tell where the package came from. E.g. if it's the "main package" of the repository you're building, it would be `pkg:npm/some-package@1.2.3?vcs_url=https://github.com/some-org/some-package@abcdef1234...`

If `some-package` doesn't have a version, using `abcdef1234...` as the version won't really achieve anything. We would have to know what `<insert popular CVE databases>` do when packages don't have versions

Reply:
Great point. I forgot that we'd have the `vcs_url` part in the purl, that will make very clear what the actual package is, so this solves my initial concern. As for the CVE part, I guess you're right, it'll be dependend on packages that need to have versions.